xmake/rules/c++/modules/xmake.lua

```diff
@@ -21,45 +21,153 @@
 -- define rule: c++.build.modules
 rule("c++.build.modules")
     set_extensions(".mpp", ".mxx", ".cppm", ".ixx")
+    add_deps("c++.build.modules.dependencies")
+    add_deps("c++.build.modules.builder")
+    add_deps("c++.build.modules.install")
+
     on_config(function (target)
-        -- we disable to build across targets in parallel, because the source files may depend on other target modules
-        -- @see https://github.com/xmake-io/xmake/issues/1858
-        local target_with_modules
+        local target_with_modules = target:sourcebatches()["c++.build.modules"] and true or false
+
         for _, dep in ipairs(target:orderdeps()) do
-            local sourcebatches = dep:sourcebatches()
-            if sourcebatches and sourcebatches["c++.build.modules"] then
-                target_with_modules = true
-                break
-            end
+            target_with_modules = dep:sourcebatches()["c++.build.modules"] and true or target_with_modules
```

Comment:
This will traverse all deps and we should try to traverse as few deps as possible.

we can add a helper function to improve it.

```lua

-- this target contains modules?
function contains_modules(target)
    local target_with_modules
    for _, dep in ipairs(target:orderdeps()) do
        local sourcebatches = dep:sourcebatches()
        if sourcebatches and sourcebatches["c++.build.modules"] then
            target_with_modules = true
            break
        end
    end
    return target_with_modules
end
```

```lua
        if module_utils.contains_modules(target) then
        end
```